src/nodes/dir.rs

```diff
@@ -247,6 +247,27 @@ impl Dir {
         }
         Dir::new_empty(ids.next(), Some(self), now)
     }
+
+    /// Same as `getattr` but with the node already locked.
+    fn getattr_unlocked(inode: u64, state: &mut MutableDir) -> NodeResult<fuse::FileAttr> {
+        let new_attr = match state.underlying_path.as_ref() {
```

Comment:
Rust can partition borrows across fields:

```rust
        if let Some(path) = &state.underlying_path {
            let fs_attr = fs::symlink_metadata(path)?;
            if !fs_attr.is_dir() {
                warn!("Path {:?} backing a directory node is no longer a directory; got {:?}",
                    path, fs_attr.file_type());
                return Err(KernelError::from_errno(errno::Errno::EIO));
            }
            state.attr = conv::attr_fs_to_fuse(path, inode, &fs_attr);
        };
```

Reply:
Excellent. That was my expectation (that borrowing `underlying_path` shouldn't have an effect on `attr`)... but I can't quite understand why your simpler rewrite makes this work.

AFAICT the critical difference is between my previous `state.underlying_path.as_ref()` and your `&state.underlying_path`, right? Both borrows are on the field...

Reply:
I don't think that's the critical difference, but I can't see the old code to check any more.

Reply:
The old code did this:

```rust
        let new_attr = match state.underlying_path.as_ref() {
            Some(path) => {
                let fs_attr = fs::symlink_metadata(path)?;
                if !fs_attr.is_dir() {
                    warn!("Path {:?} backing a directory node is no longer a directory; got {:?}",
                        path, fs_attr.file_type());
                    return Err(KernelError::from_errno(errno::Errno::EIO));
                }
                Some(conv::attr_fs_to_fuse(path, inode, &fs_attr))
            },
            None => None,
        };
        if let Some(new_attr) = new_attr {
            state.attr = new_attr;
        }
```

(This time around note that I pushed review changes as new commits instead of amending the previous ones, precisely to avoid things vanishing during review time. I could revive these in the UI by clicking on the diff for the first commit. I've been told that's how GitHub PRs should be used... which is annoying because it doesn't work for commit chains.)